File: silicon/parser.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import *
from copy import copy

from silicon.lexer import Token, TokenKind
from silicon import ast
from silicon.diagnostics import *
from silicon.source import *


def parse_tokens(tokens: List[Token]) -> ast.Root:
    p = Parser(tokens=tokens, last_loc=tokens[0].loc)
    return parse_root(p)


@dataclass
class Parser:
    tokens: List[Token]
    last_loc: Loc

    def loc(self) -> Loc:
        return self.tokens[0].loc

    def consume(self) -> Token:
        t = self.tokens[0]
        self.tokens = self.tokens[1:]
        self.last_loc = t.loc
        return t

    def consume_if(self, kind: TokenKind) -> Optional[Token]:
        if self.tokens[0].kind == kind:
            return self.consume()
        return None

    def require(self, kind: TokenKind, msg: Optional[str] = None) -> Token:
        if token := self.consume_if(kind):
            return token
        msg = msg or kind.name
        emit_error(self.loc(),
                   f"expected {msg}, found {self.tokens[0].kind.name}")

    def isa(self, kind: TokenKind) -> bool:
        return self.tokens[0].kind == kind

    def not_delim(self, *args: TokenKind) -> bool:
        return self.tokens[0].kind not in (TokenKind.EOF, *args)


def parse_root(p: Parser) -> ast.Root:
    loc = p.loc()
    items: List[ast.Item] = []
    while not p.isa(TokenKind.EOF):
        items.append(parse_item(p))
    return ast.Root(loc=loc | p.last_loc, items=items)


def parse_item(p: Parser) -> ast.Item:
    # Parse module definitions.
    if kw := p.consume_if(TokenKind.KW_MOD):
        name = p.require(TokenKind.IDENT, "module name")
        p.require(TokenKind.LCURLY)
        stmts: List[ast.Stmt] = []
        while p.not_delim(TokenKind.RCURLY):
            stmts.append(parse_stmt(p))
        p.require(TokenKind.RCURLY)
        return ast.ModItem(loc=name.loc,
                           full_loc=kw.loc | p.last_loc,
                           name=name,
                           stmts=stmts)

    # Parse function definitions.
    if kw := p.consume_if(TokenKind.KW_FN):
        name = p.require(TokenKind.IDENT, "function name")

        # Parse arguments.
        p.require(TokenKind.LPAREN)
        args: List[ast.FnArg] = []
        while p.not_delim(TokenKind.RPAREN):
            arg_name = p.require(TokenKind.IDENT, "argument name")
            p.require(TokenKind.COLON)
            arg_type = parse_type(p)
            args.append(
                ast.FnArg(loc=arg_name.loc,
                          full_loc=arg_name.loc | p.last_loc,
                          name=arg_name,
                          ty=arg_type))
            if not p.consume_if(TokenKind.COMMA):
                break
        p.require(TokenKind.RPAREN)

        # Parse return type.
        return_ty: Optional[ast.AstType] = None
        if p.consume_if(TokenKind.ARROW):
            return_ty = parse_type(p)

        # Parse body.
        p.require(TokenKind.LCURLY)
        stmts = []
        while p.not_delim(TokenKind.RCURLY):
            stmts.append(parse_stmt(p))
        p.require(TokenKind.RCURLY)
        return ast.FnItem(loc=name.loc,
                          full_loc=kw.loc | p.last_loc,
                          name=name,
                          args=args,
                          return_ty=return_ty,
                          stmts=stmts)

    emit_error(p.loc(), f"expected item, found {p.tokens[0].kind.name}")


def parse_stmt_or_expr(p: Parser) -> ast.Stmt | ast.Expr:
    loc = p.loc()

    # Parse stray semicolons.
    if p.consume_if(TokenKind.SEMICOLON):
        return ast.EmptyStmt(loc=loc | p.last_loc, full_loc=loc | p.last_loc)

    # Parse input port declarations.
    if kw := p.consume_if(TokenKind.KW_INPUT):
        name = p.require(TokenKind.IDENT, "input name")
        p.require(TokenKind.COLON)
        ty = parse_type(p)
        p.require(TokenKind.SEMICOLON)
        return ast.InputStmt(loc=name.loc,
                             full_loc=loc | p.last_loc,
                             name=name,
                             ty=ty)

    # Parse output port declarations.
    if kw := p.consume_if(TokenKind.KW_OUTPUT):
        name = p.require(TokenKind.IDENT, "output name")
        p.require(TokenKind.COLON)
        ty = parse_type(p)
        maybeExpr: Optional[ast.Expr] = None
        if assign := p.consume_if(TokenKind.ASSIGN):
            maybeExpr = parse_expr(p)
        p.require(TokenKind.SEMICOLON)
        return ast.OutputStmt(loc=name.loc,
                              full_loc=loc | p.last_loc,
                              name=name,
                              ty=ty,
                              expr=maybeExpr)

    # Parse let bindings.
    if kw := p.consume_if(TokenKind.KW_LET):
        name = p.require(TokenKind.IDENT, "output name")
        maybeTy = None
        if colon := p.consume_if(TokenKind.COLON):
            maybeTy = parse_type(p)
        maybeExpr = None
        if assign := p.consume_if(TokenKind.ASSIGN):
            maybeExpr = parse_expr(p)
        p.require(TokenKind.SEMICOLON)
        return ast.LetStmt(loc=name.loc,
                           full_loc=loc | p.last_loc,
                           name=name,
                           ty=maybeTy,
                           expr=maybeExpr)

    # Parse return statements.
    if kw := p.consume_if(TokenKind.KW_RETURN):
        maybeExpr = None
        if not p.isa(TokenKind.SEMICOLON):
            maybeExpr = parse_expr(p)
        p.require(TokenKind.SEMICOLON)
        return ast.ReturnStmt(loc=kw.loc,
                              full_loc=loc | p.last_loc,
                              expr=maybeExpr)

    # Otherwise this is a statement that starts with an expression.
    expr = parse_expr(p)
    if p.consume_if(TokenKind.SEMICOLON):
        return ast.ExprStmt(loc=expr.loc, full_loc=loc | p.last_loc, expr=expr)

    # Parse assignments.
    if assign := p.consume_if(TokenKind.ASSIGN):
        rhs = parse_expr(p)
        p.require(TokenKind.SEMICOLON)
        return ast.AssignStmt(loc=assign.loc,
                              full_loc=loc | p.last_loc,
                              lhs=expr,
                              rhs=rhs)

    return expr


def parse_stmt(p: Parser) -> ast.Stmt:
    stmt = parse_stmt_or_expr(p)
    if isinstance(stmt, ast.Stmt):
        return stmt
    if expr_requires_semicolon(stmt):
        p.require(TokenKind.SEMICOLON)
    return ast.ExprStmt(loc=stmt.loc,
                        full_loc=stmt.loc | p.last_loc,
                        expr=stmt)


# Check whether an expression requires a trailing semicolon if used as a
# statement in a block.
def expr_requires_semicolon(expr: ast.Expr) -> bool:
    return not isinstance(expr, ast.BlockExpr)


def parse_type(p: Parser) -> ast.AstType:
    loc = p.loc()

    # Parse tuples or the `()` unit type.
    if p.consume_if(TokenKind.LPAREN):
        if p.consume_if(TokenKind.RPAREN):
            return ast.UnitType(loc=loc | p.last_loc)
        fields: List[ast.AstType] = []
        while p.not_delim(TokenKind.RPAREN):
            fields.append(parse_type(p))
            if not p.consume_if(TokenKind.COMMA):
                break
        p.require(TokenKind.RPAREN)
        if len(fields) < 2:
            emit_error(loc | p.last_loc, "tuple requires at least two fields")
        return ast.TupleType(loc=loc | p.last_loc, fields=fields)

    # Parse types that start with a name.
    if p.isa(TokenKind.IDENT):
        name = p.tokens[0].spelling()
        if name == "uint":
            token = p.consume()
            p.require(TokenKind.LT)
            size = p.require(TokenKind.NUM_LIT)
            try:
                size_value = int(size.spelling())
            except:
                emit_error(size.loc, "invalid size for `uint`")
            p.require(TokenKind.GT)
            return ast.UIntType(loc=loc | p.last_loc,
                                size=size_value,
                                size_loc=size.loc)

        if name == "Wire":
            token = p.consume()
            p.require(TokenKind.LT)
            inner = parse_type(p)
            p.require(TokenKind.GT)
            return ast.WireType(loc=loc | p.last_loc, inner=inner)

        if name == "Reg":
            token = p.consume()
            p.require(TokenKind.LT)
            inner = parse_type(p)
            p.require(TokenKind.GT)
            return ast.RegType(loc=loc | p.last_loc, inner=inner)

    emit_error(p.loc(), f"expected type, found {p.tokens[0].kind.name}")


def parse_expr(
    p: Parser,
    min_prec: ast.Precedence = ast.Precedence.MIN,
) -> ast.Expr:
    expr = parse_prefix_expr(p)
    while infix := parse_infix_expr(p, expr, min_prec):
        expr = infix
    return expr


def parse_prefix_expr(p: Parser) -> ast.Expr:
    loc = p.loc()

    # Parse unary operators.
    if op := ast.UNARY_OPS.get(p.tokens[0].kind):
        op_loc = p.consume().loc
        arg = parse_prefix_expr(p)
        return ast.UnaryExpr(
            loc=op_loc,
            full_loc=loc | arg.loc,
            op=op,
            arg=arg,
        )

    # Otherwise parse a primary expression and optional suffix expressions.
    expr = parse_primary_expr(p)
    while suffix := parse_suffix_expr(p, expr):
        expr = suffix
    return expr


def parse_infix_expr(
    p: Parser,
    lhs: ast.Expr,
    min_prec: ast.Precedence,
) -> Optional[ast.Expr]:
    # Parse binary operators.
    if op := ast.BINARY_OPS.get(p.tokens[0].kind):
        prec = ast.BINARY_PRECEDENCE[op]
        if prec > min_prec:
            op_loc = p.consume().loc
            rhs = parse_expr(p, prec)
            return ast.BinaryExpr(
                loc=op_loc,
                full_loc=lhs.loc | rhs.loc,
                op=op,
                lhs=lhs,
                rhs=rhs,
            )

    return None


def parse_suffix_expr(p: Parser, expr: ast.Expr) -> Optional[ast.Expr]:
    # Parse field calls, such as `a.b(c)`, or tuple field accesses.
    if dot := p.consume_if(TokenKind.DOT):
        # a.b(c)
        if name := p.consume_if(TokenKind.IDENT):
            p.require(TokenKind.LPAREN)
            args: List[ast.Expr] = []
            while p.not_delim(TokenKind.RPAREN):
                args.append(parse_expr(p))
                if not p.consume_if(TokenKind.COMMA):
                    break
            p.require(TokenKind.RPAREN)
            return ast.FieldCallExpr(
                loc=name.loc,
                full_loc=expr.loc | p.last_loc,
                target=expr,
                name=name,
                args=args,
            )

        # a.42
        if index := p.consume_if(TokenKind.NUM_LIT):
            try:
                index_value = int(index.spelling())
            except:
                emit_error(index.loc, "invalid tuple index")
            return ast.TupleFieldExpr(loc=index.loc,
                                      full_loc=expr.loc | p.last_loc,
                                      target=expr,
                                      field=index_value)

        emit_error(dot.loc,
                   f"expected field name, found {p.tokens[0].kind.name}")

    return None


def parse_primary_expr(p: Parser) -> ast.Expr:
    loc = p.loc()

    # Parse number literals.
    if lit := p.consume_if(TokenKind.NUM_LIT):
        m = re.match(r'(\d+)(u(\d+))?', lit.spelling())
        if not m:
            emit_error(
                lit.loc,
                f"expected number literal of the form `<V>` or `<V>u<N>`")
        return ast.IntLitExpr(
            loc=lit.loc,
            full_loc=lit.loc,
            value=int(m[1]),
            width=int(m[3]) if m[2] else None,
        )

    # Parse identifiers and calls.
    if token := p.consume_if(TokenKind.IDENT):
        if paren := p.consume_if(TokenKind.LPAREN):
            args: List[ast.Expr] = []
            while p.not_delim(TokenKind.RPAREN):
                args.append(parse_expr(p))
                if not p.consume_if(TokenKind.COMMA):
                    break
            p.require(TokenKind.RPAREN)
            return ast.CallExpr(
                loc=token.loc,
                full_loc=token.loc | p.last_loc,
                name=token,
                args=args,
            )

        return ast.IdentExpr(loc=token.loc, full_loc=token.loc, name=token)

    # Parse parenthesized expressions, tuples, or `()` unit literals.
    if p.consume_if(TokenKind.LPAREN):
        if p.consume_if(TokenKind.RPAREN):
            return ast.UnitLitExpr(loc=loc | p.last_loc,
                                   full_loc=loc | p.last_loc)
        fields: List[ast.Expr] = []
        while p.not_delim(TokenKind.RPAREN):
            fields.append(parse_expr(p))
            if not p.consume_if(TokenKind.COMMA):
                break
        p.require(TokenKind.RPAREN)
        if len(fields) == 1:
            return ast.ParenExpr(loc=fields[0].loc,
                                 full_loc=loc | p.last_loc,
                                 expr=fields[0])
        else:
            return ast.TupleExpr(loc=loc | p.last_loc,
                                 full_loc=loc | p.last_loc,
                                 fields=fields)

    # Parse blocks.
    if p.isa(TokenKind.LCURLY):
        return parse_block_expr(p)

    emit_error(p.loc(), f"expected expression, found {p.tokens[0].kind.name}")


def parse_block_expr(p: Parser) -> ast.BlockExpr:
    loc = p.loc()
    stmts: List[ast.Stmt] = []
    result: Optional[ast.Expr] = None
    p.require(TokenKind.LCURLY)

    while p.not_delim(TokenKind.RCURLY):
        stmt = parse_stmt_or_expr(p)

        # If the statement was an expression, this may be the final expression
        # in the block which defines the block's result value.
        if isinstance(stmt, ast.Expr):
            # If this was an expression without subsequent semicolon, and the
            # next token is the closing brace `}`, this expression is the result
            # value of the block.
            if p.isa(TokenKind.RCURLY):
                result = stmt
                break

            # Otherwise the block isn't done yet. Complain about the missing
            # semicolon.
            if expr_requires_semicolon(stmt):
                after = copy(p.last_loc)
                after.offset += after.length
                after.length = 0
                emit_error(after, "expected `;` after expression")

            # Wrap this expression up as a statement, since there are more left
            # in the block.
            stmt = ast.ExprStmt(loc=stmt.loc,
                                full_loc=stmt.full_loc,
                                expr=stmt)

        stmts.append(stmt)

    p.require(TokenKind.RCURLY)
    return ast.BlockExpr(loc=loc | p.last_loc,
                         full_loc=loc | p.last_loc,
                         stmts=stmts,
                         result=result)

File: lib/Syntax/Parser.cpp
//===----------------------------------------------------------------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//

#include "silicon/Syntax/AST.h"
#include "silicon/Syntax/Parser.h"
#include "silicon/Syntax/Tokens.h"
#include "llvm/ADT/StringExtras.h"

using namespace silicon;

//===----------------------------------------------------------------------===//
// Parser Infrastructure
//===----------------------------------------------------------------------===//

Parser::Parser(Lexer &lexer, AST &ast) : lexer(lexer), ast(ast) {
  token = lexer.next();
}

Token Parser::consume() {
  auto consumedToken = token;
  token = lexer.next();
  return consumedToken;
}

Token Parser::consumeIf(TokenKind kind) {
  if (isa(kind))
    return consume();
  return {token.spelling.substr(0, 0), TokenKind::Eof};
}

Token Parser::require(TokenKind kind, const Twine &msg) {
  if (isa(kind))
    return consume();

  // Only emit an error if the lexer has not already emitted one.
  if (!token.isError()) {
    auto d = mlir::emitError(loc(), "expected ");
    if (msg.isTriviallyEmpty())
      d << symbolizeTokenKind(kind);
    else
      d << msg;
    d << ", found " << token;
  }

  return {token.spelling.substr(0, 0), TokenKind::Eof};
}

//===----------------------------------------------------------------------===//
// Grammar
//===----------------------------------------------------------------------===//

ast::Root *Parser::parseRoot() {
  SmallVector<ast::Item *> items;
  while (token) {
    auto *item = parseItem();
    if (!item)
      return {};
    items.push_back(item);
  }
  return ast.create<ast::Root>({ast.array(items)});
}

ast::Item *Parser::parseItem() {
  // Parse function definitions.
  if (auto kw = consumeIf(TokenKind::Kw_fn))
    return parseFnItem(kw);

  // If we get here we didn't find a keyword that starts an item.
  if (!token.isError())
    mlir::emitError(loc()) << "expected item, found " << token;
  return {};
}

//===----------------------------------------------------------------------===//
// Functions
//===----------------------------------------------------------------------===//

ast::FnItem *Parser::parseFnItem(Token kw) {
  // Parse the function name.
  auto name = require(TokenKind::Ident, "function name");
  if (!name)
    return {};

  // Parse the function arguments.
  SmallVector<ast::FnArg *> args;
  if (!require(TokenKind::LParen))
    return {};
  while (notAtDelimiter(TokenKind::RParen)) {
    auto *arg = parseFnArg();
    if (!arg)
      return {};
    args.push_back(arg);
    if (!consumeIf(TokenKind::Comma))
      break;
  }
  if (!require(TokenKind::RParen))
    return {};

  // Parse the optional function return type.
  ast::Type *returnType = {};
  if (consumeIf(TokenKind::Arrow)) {
    returnType = parseType();
    if (!returnType)
      return {};
  }

  // Parse the function body.
  auto *body = parseBlockExpr();
  if (!body)
    return {};

  return ast.create<ast::FnItem>({{ast::ItemKind::Fn, loc(name)},
                                  name.spelling,
                                  ast.array(args),
                                  returnType,
                                  body});
}

ast::FnArg *Parser::parseFnArg() {
  // Parse the argument name.
  auto name = require(TokenKind::Ident, "argument name");
  if (!name)
    return {};

  // Parse the argument type.
  if (!require(TokenKind::Colon))
    return {};
  auto *type = parseType();
  if (!type)
    return {};

  return ast.create<ast::FnArg>({loc(name), name.spelling, type});
}

//===----------------------------------------------------------------------===//
// Types
//===----------------------------------------------------------------------===//

// NOLINTNEXTLINE(misc-no-recursion)
ast::Type *Parser::parseType() {
  // Parse the `const` type.
  if (auto kw = consumeIf(TokenKind::Kw_const)) {
    auto *type = parseType();
    if (!type)
      return {};
    return ast.create<ast::ConstType>({{ast::TypeKind::Const, loc(kw)}, type});
  }

  // Parse the `int` type.
  if (auto kw = consumeIf(TokenKind::Kw_int))
    return ast.create<ast::Type>({ast::TypeKind::Int, loc(kw)});

  // Parse the `uint` type.
  if (auto kw = consumeIf(TokenKind::Kw_uint)) {
    if (!require(TokenKind::Lt))
      return {};
    // Use relational precedence to parse the width expression to not consume
    // the closing `>`.
    auto *width = parseExpr(ast::Precedence::Rel);
    if (!width)
      return {};
    if (!require(TokenKind::Gt))
      return {};
    return ast.create<ast::UIntType>({{ast::TypeKind::UInt, loc(kw)}, width});
  }

  // If we get here we didn't find a keyword that starts a type.
  if (!token.isError())
    mlir::emitError(loc()) << "expected type, found " << token;
  return {};
}

//===----------------------------------------------------------------------===//
// Expressions
//===----------------------------------------------------------------------===//

/// Check if an expression requires a semicolon when used as a statement.
static bool requiresSemicolon(ast::Expr *expr) {
  return !isa<ast::BlockExpr, ast::IfExpr>(expr);
}

/// Check whether the given token kind is a unary operator and return the
/// corresponding AST unary operator.
static std::optional<ast::UnaryOp> getUnaryOp(TokenKind kind) {
  switch (kind) {
#define AST_UNARY(NAME, TOKEN)                                                 \
  case TokenKind::TOKEN:                                                       \
    return ast::UnaryOp::NAME;
#include "silicon/Syntax/AST.def"
  default:
    return {};
  }
}

/// Check whether the given token kind is a binary operator and return the
/// corresponding AST binary operator.
static std::optional<ast::BinaryOp> getBinaryOp(TokenKind kind) {
  switch (kind) {
#define AST_BINARY(NAME, TOKEN, PREC)                                          \
  case TokenKind::TOKEN:                                                       \
    return ast::BinaryOp::NAME;
#include "silicon/Syntax/AST.def"
  default:
    return {};
  }
}

// NOLINTNEXTLINE(misc-no-recursion)
ast::Expr *Parser::parseExpr(ast::Precedence minPrec) {
  // Parse a primary expression first. This will form the LHS of any operators
  // that may follow.
  auto *lhs = parsePrimaryExpr();
  if (!lhs)
    return {};

  // Parse any operators that may follow the primary expression.
  while (token) {
    // Parse function calls.
    if (isa(TokenKind::LParen) && ast::Precedence::Suffix > minPrec) {
      auto lparen = consume();
      SmallVector<ast::Expr *> args;
      while (notAtDelimiter(TokenKind::RParen)) {
        auto *arg = parseExpr();
        if (!arg)
          return {};
        args.push_back(arg);
        if (!consumeIf(TokenKind::Comma))
          break;
      }
      if (!require(TokenKind::RParen))
        return {};
      lhs = ast.create<ast::CallExpr>(
          {{ast::ExprKind::Call, loc(lparen)}, lhs, ast.array(args)});
      continue;
    }

    // Parse indexing expressions.
    if (isa(TokenKind::LBrack) && ast::Precedence::Suffix > minPrec) {
      auto lbrack = consume();
      auto *index = parseExpr();
      if (!index)
        return {};
      if (!require(TokenKind::RBrack))
        return {};
      lhs = ast.create<ast::IndexExpr>(
          {{ast::ExprKind::Index, loc(lbrack)}, lhs, index});
      continue;
    }

    // Parse binary operators.
    if (auto op = getBinaryOp(token.kind)) {
      auto prec = ast::getPrecedence(*op);
      if (prec > minPrec) {
        auto opToken = consume();
        auto *rhs = parseExpr(prec);
        if (!rhs)
          return {};
        lhs = ast.create<ast::BinaryExpr>(
            {{ast::ExprKind::Binary, loc(opToken)}, *op, lhs, rhs});
        continue;
      }
    }

    // If we get here we didn't recognize the token as part of an expression.
    break;
  }

  return lhs;
}

// NOLINTNEXTLINE(misc-no-recursion)
ast::Expr *Parser::parsePrimaryExpr() {
  // Parse identifiers.
  if (auto ident = consumeIf(TokenKind::Ident))
    return ast.create<ast::IdentExpr>(
        {{ast::ExprKind::Ident, loc(ident)}, ident.spelling});

  // Parse number literals.
  if (auto lit = consumeIf(TokenKind::NumLit))
    return parseNumberLiteral(lit);

  // Parse block expressions.
  if (isa(TokenKind::LCurly))
    return parseBlockExpr();

  // Parse if expressions.
  if (auto kw = consumeIf(TokenKind::Kw_if)) {
    auto *condition = parseExpr(ast::Precedence::Min);
    if (!condition)
      return {};
    auto *thenExpr = parseBlockExpr();
    if (!thenExpr)
      return {};
    ast::Expr *elseExpr = {};
    if (consumeIf(TokenKind::Kw_else)) {
      if (isa(TokenKind::Kw_if))
        elseExpr = parseExpr(ast::Precedence::Min);
      else
        elseExpr = parseBlockExpr();
      if (!elseExpr)
        return {};
    }
    return ast.create<ast::IfExpr>(
        {{ast::ExprKind::If, loc(kw)}, condition, thenExpr, elseExpr});
  }

  // Parse return expressions.
  if (auto kw = consumeIf(TokenKind::Kw_return)) {
    ast::Expr *value = {};
    if (!isa(TokenKind::Semicolon, TokenKind::Comma, TokenKind::RParen,
             TokenKind::RBrack, TokenKind::RCurly)) {
      value = parseExpr(ast::Precedence::Min);
      if (!value)
        return {};
    }
    return ast.create<ast::ReturnExpr>(
        {{ast::ExprKind::Return, loc(kw)}, value});
  }

  // Parse constant expressions.
  if (auto kw = consumeIf(TokenKind::Kw_const)) {
    auto *value = parseBlockExpr();
    if (!value)
      return {};
    return ast.create<ast::ConstExpr>({{ast::ExprKind::Const, loc(kw)}, value});
  }

  // Parse unary operators.
  if (auto op = getUnaryOp(token.kind)) {
    auto opToken = consume();
    auto *arg = parseExpr(ast::Precedence::Prefix);
    if (!arg)
      return {};
    return ast.create<ast::UnaryExpr>(
        {{ast::ExprKind::Unary, loc(opToken)}, *op, arg});
  }

  // If we get here we didn't find anything that looks like an expression.
  if (!token.isError())
    mlir::emitError(loc()) << "expected expression, found " << token;
  return {};
}

/// Check whether the given character is a valid digit for the given base.
static bool isValidDigitForBase(char c, int base) {
  if (c >= '0' && c <= '9')
    return (c - '0') < base;
  c = llvm::toLower(c);
  if (c >= 'a' && c <= 'f')
    return (c - 'a' + 10) < base;
  return false;
}

ast::NumLitExpr *Parser::parseNumberLiteral(Token lit) {
  StringRef spelling = lit.spelling;

  // Determine the base.
  unsigned base = 10;
  if (spelling.consume_front("0b"))
    base = 2;
  else if (spelling.consume_front("0o"))
    base = 8;
  else if (spelling.consume_front("0x"))
    base = 16;

  // Filter out `_` and check for invalid digits for the given base.
  SmallString<32> digits;
  digits.reserve(spelling.size());
  for (unsigned i = 0, e = spelling.size(); i != e; ++i) {
    if (spelling[i] == '_')
      continue;
    if (!isValidDigitForBase(spelling[i], base)) {
      mlir::emitError(loc(spelling.substr(i)))
          << "`" << spelling[i] << "` is not a valid base-" << base << " digit";
      return {};
    }
    digits.push_back(spelling[i]);
  }

  // Parse the integer. At this point we know that all characters in `digits`
  // are valid digits for the given base.
  APInt value;
  auto hasRemainder = digits.str().getAsInteger(base, value);
  assert(!hasRemainder);

  // Resize the APInt to the smallest possible width that can hold the value.
  value = value.zextOrTrunc(value.getActiveBits());

  return ast.create<ast::NumLitExpr>(
      {{ast::ExprKind::NumLit, loc(lit)}, value});
}

// NOLINTNEXTLINE(misc-no-recursion)
ast::BlockExpr *Parser::parseBlockExpr() {
  auto lcurly = require(TokenKind::LCurly);
  if (!lcurly)
    return {};

  SmallVector<ast::Stmt *> stmts;
  ast::Expr *result = {};

  // Parse the statements in the block. The very last statement in the block may
  // be an expression without a terminating semicolon, in which case that
  // expressions becomes the block's return value.
  while (notAtDelimiter(TokenKind::RCurly)) {
    // Parse an expression or statement.
    auto stmtOrExpr = parseStmtOrExpr();
    if (!stmtOrExpr)
      return {};

    // If this is an expression, this may be the final expression in the block
    // which defines the block's return value.
    if (auto *expr = dyn_cast<ast::Expr *>(stmtOrExpr)) {
      // If the next token is a closing curly brace, use the expression as the
      // block's return value and break out of the loop.
      if (isa(TokenKind::RCurly)) {
        result = expr;
        break;
      }

      // If the expression requires a semicolon when used as a statement, which
      // is true for almost all expressions, consume that semicolon.
      if (requiresSemicolon(expr))
        if (!require(TokenKind::Semicolon))
          return {};

      // Wrap the expression in a statement so we can add it to the block.
      stmts.push_back(
          ast.create<ast::ExprStmt>({{ast::StmtKind::Expr, expr->loc}, expr}));
      continue;
    }

    // Since this was not an expression, it must be a statement.
    stmts.push_back(cast<ast::Stmt *>(stmtOrExpr));
  }

  // Consume the closing curly brace.
  if (!require(TokenKind::RCurly))
    return {};

  return ast.create<ast::BlockExpr>(
      {{ast::ExprKind::Block, loc(lcurly)}, stmts, result});
}

//===----------------------------------------------------------------------===//
// Statements
//===----------------------------------------------------------------------===//

// NOLINTNEXTLINE(misc-no-recursion)
Parser::ExprOrStmt Parser::parseStmtOrExpr() {
  // Ignore stray semicolons.
  if (auto token = consumeIf(TokenKind::Semicolon))
    return ast.create<ast::Stmt>({ast::StmtKind::Empty, loc(token)});

  // Parse let bindings.
  if (auto kw = consumeIf(TokenKind::Kw_let)) {
    // Parse the variable name.
    auto name = require(TokenKind::Ident, "variable name");
    if (!name)
      return {};

    // Parse the optional type of the variable.
    ast::Type *type = {};
    if (consumeIf(TokenKind::Colon)) {
      type = parseType();
      if (!type)
        return {};
    }

    // Parse the optional value of the variable.
    ast::Expr *value = {};
    if (consumeIf(TokenKind::Assign)) {
      value = parseExpr(ast::Precedence::Min);
      if (!value)
        return {};
    }

    if (!require(TokenKind::Semicolon))
      return {};
    return ast.create<ast::LetStmt>(
        {{ast::StmtKind::Let, loc(kw)}, name.spelling, type, value});
  }

  // Otherwise this is an expression.
  return parseExpr();
}
